### PR TITLE
Add Maven profile for revapi API compatibility checks

### PR DESCRIPTION
**Pull request by spannm**
Revapi and revapi-maven-plugin can be found here: https://revapi.org/

- Main revapi configuration and versions defined in `jdbi3-build-parent`
- Maven profile 'revapi' defined in `jdbi3-build-parent`, goal `revapi:check`
- Maven profile 'revapi-report' defined in jdbi3-build-parent, goal revapi:report
- Modules not to be checked may set property `jdbi.check.skip-revapi=true` (e.g. modules for internal or testing use)
- Modules for which violations should not fail the build should set property `jdbi.check.fail-revapi=false`

Revapi compares the project to the most recent release (e.g. 3.37.1-SNAPSHOT to 3.37.0) looking for violations of semantic versioning rules. The ruleset can be relaxed by configuration.

Violations found by revapi can be explicitly ignored.
I've done this in modules `core` and `freemarker` by example (they would have otherwise failed, would revert this of course before merging into master). **Update:** Configuration now rolled back

To other failing modules I've added `jdbi.check.fail-revapi=false`.  **Update:** Configuration now rolled back

To see the problems look for 'revapi' in the build output, in particular for string 'API problems found' and onwards (examples below).

## Maven command line
`mvn clean install -P revapi,fast -D no-docker`

This takes a little over 2 minutes on my lame Ubuntu laptop as opposed to 1:20 min without revapi. So it does add overhead but less than one would expect.

## Maven command line to create and view revapi reports
```
mvn clean install -P revapi-report,fast -D no-docker
find . -name revapi-report.html -exec firefox {} \;
```

## Build output / no compatibility problems
```
[INFO] --- revapi-maven-plugin:0.15.0:check (default) @ jdbi3-core ---
Downloading from nexus: http://nexus-swe.baag:8081/nexus/content/groups/public/org/jdbi/jdbi3-core/maven-metadata.xml
Downloaded from nexus: http://nexus-swe.baag:8081/nexus/content/groups/public/org/jdbi/jdbi3-core/maven-metadata.xml (3.0 kB at 5.7 kB/s)
[INFO] Comparing [org.jdbi:jdbi3-core:jar:3.37.0] against [org.jdbi:jdbi3-core:jar:3.37.1-SNAPSHOT] (including their transitive dependencies).
[INFO] API checks completed without failures.
```

## Build output / with compatibility problems

```
[INFO] --- revapi-maven-plugin:0.15.0:check (default) @ jdbi3-jpa ---
Downloading from nexus: http://nexus-swe.baag:8081/nexus/content/groups/public/org/jdbi/jdbi3-jpa/maven-metadata.xml
Downloaded from nexus: http://nexus-swe.baag:8081/nexus/content/groups/public/org/jdbi/jdbi3-jpa/maven-metadata.xml (3.6 kB at 10 kB/s)
[INFO] Comparing [org.jdbi:jdbi3-jpa:jar:3.37.0] against [org.jdbi:jdbi3-jpa:jar:3.37.1-SNAPSHOT] (including their transitive dependencies).
[INFO] API problems found.
[INFO] If you're using the semver-ignore extension, update your module's version to one compatible with the current changes (e.g. mvn package revapi:update-versions). If you want to explicitly ignore these changes or provide justifications for them, add the xml snippets to your Revapi configuration for the "revapi.differences" extension.

```

### CODE REVIEW

File: json/pom.xml [hunk 1]
```diff
@@ -28,6 +28,7 @@
 
     <properties>
         <moduleName>org.jdbi.v3.json</moduleName>
+        <jdbi.check.fail-revapi>false</jdbi.check.fail-revapi>
```
Comment (hgschmie) [inline json/pom.xml:31]: wrong order?

**Review by hgschmie — CHANGES_REQUESTED** [review]
I think my biggest gripe with this is that this is big and contains a lot of things we should not care about.

- we should not turn testing on or off for separate modules. We should test all modules
- there are actual ignored differences in the core pom enumerated, this adds 70 lines for a single plugin and I assume that is not the end. THis makes me very uncomfortable. Why is that needed?
- there are freemarker classes in the freemarker pom. Can't we just limit to org.jdbi and ignore changes in everything else?

This is a pretty big PR for a basic function of "pull the previous version, run a diff". I would have expected something along the lines of "here is a bit of config for a plugin, check only org.jdbi, ignore everything that has *.internal.* or *.inlined.* in the package name, create a report". If that report shows problems, that is ok. We know we made mistakes in the past, this is intended to help us to get better.

**Comment by sman-81** [comment]
Thanks for your feedback @hgschmie 

>     * we should not turn testing on or off for separate modules. We should test all modules

What is the benefit from having revapi checks on internal and testing modules?
Noone would be interested in the findings. So they can be skipped.

>     * there are actual ignored differences in the core pom enumerated, this adds 70 lines for a single plugin and I assume that is not the end. THis makes me very uncomfortable. Why is that needed?

This is for demonstration only and will never get into master. Such ignored differences become obsolete after the release such changed API. Please see the initial comments of my PR above.

>     * there are freemarker classes in the freemarker pom. Can't we just limit to org.jdbi and ignore changes in everything else?

That's worth investigating and should be supported by the plugin.

> This is a pretty big PR for a basic function of "pull the previous version, run a diff".

The plugin is not part of the build cycle but is defined as Maven profile. I don't think its that intrusive :)
Since Maven site not something the project seems to use, I chose `revapi:check` over a site report (`revapi:report`).

3.37.0 -> 3.37.1 has a lot of API changes. So I suppose the timing of showcasing revapi just now was unfortunate.

I am still fuzzy about what we want from revapi. Following semver fully cannot be it. The major version is effectively locked at 3. If we still want to permit code changes that would require a new major according to semver, we have to relax the usage of semver and possibly also the configuration of revapi.

File: internal/build/pom.xml [hunk 2]
```diff
@@ -107,9 +109,11 @@
         <jdbi.check.fail-detekt>${jdbi.check.fail-kotlin}</jdbi.check.fail-detekt>
         <jdbi.check.fail-kotlin>${basepom.check.fail-extended}</jdbi.check.fail-kotlin>
         <jdbi.check.fail-ktlint>${jdbi.check.fail-kotlin}</jdbi.check.fail-ktlint>
+        <jdbi.check.fail-revapi>true</jdbi.check.fail-revapi>
```
[the remark this answers is not in the document]
Comment (sman-81) [reply internal/build/pom.xml:112]: That's how I had it at first. However the properties are controlled by properties in jdbi modules and the Maven profile 'revapi'. So a basepom default value will never be used.
I should remove the initialization of these properties altogether.

File: internal/pom.xml [hunk 3]
```diff
@@ -36,5 +36,6 @@
     <properties>
         <basepom.deploy.skip>true</basepom.deploy.skip>
         <basepom.install.skip>true</basepom.install.skip>
+        <jdbi.check.skip-revapi>true</jdbi.check.skip-revapi>
```
[the remark this answers is not in the document]
Comment (sman-81) [reply internal/pom.xml:39]: oddly I had revapi fail on me without that property

[the remark this answers is not in the document]
Comment (sman-81) [reply internal/pom.xml]: got it working

**Comment by sman-81** [comment]
Just rebased this PR to master again.

This PR adds revapi capabilities in a non-intrusive way (using a Maven profile to be activated in order to run checks).

We've talked about this on Slack and feedback was positive. However we have not really reached a decision.
@hgschmie @stevenschlansker Gentlemen, do we want revapi in the project or not? I am fine either way :)

**Comment by stevenschlansker** [comment]
I am in general supportive of having stronger checks, but I don't really have experience with revapi specifically to make a decision one way or the other. The build has gotten a lot more complex recently, so my ask would be that if we bring this in for the long term, we automate enough that it does not add more manual steps to the build release process.  I defer to Henning for the rest.

**Comment by hgschmie** [comment]
See also #920

**Comment by hgschmie** [comment]
We shipped japicmp as part of the 3.44.0 release, which makes this PR redundant.

**Pull request closed without merging**